Proposal to Surface Sponsorship Information in the CLI (#14992)
NuGet/Home: accepted/2026/sponsorship-cli.md

## Motivation

Sponsorship links already exist and are maintained by package owners on nuget.org — the data is real and current, it's just invisible outside the website. 
The core problem is getting existing nuget.org sponsorship data into developer tooling

This proposal adds a non-disruptive, opt-in way to surface sponsorship information, and is scoped as the first of three phases:

1. **Phase 1 (this proposal): CLI** — `dotnet list package --sponsor`.
2. **Phase 2: Visual Studio Package Manager UI** — surfacing sponsorship links using the same hyperlink pattern PM UI already uses for project/license/report-abuse URLs, directly addressing the demand in [nuget/home#14739](https://github.com/nuget/home/issues/14739).
3. **Phase 3: CLI popup link-out** — opening a nuget.org-hosted sponsorship page (`nuget.org/.../{id}/sponsor`) directly from the CLI, so developers get a richer, always-current destination without the CLI needing to hardcode or format every payment provider's link itself.

Providing sponsor links from nuget.org to the client is the key motivation for this proposal. 

### Why not restore?

> Restore runs on every build; adding informational (non-actionable) content there, even opt-in, risks alarm fatigue and encourages developers to disable the feature entirely rather than engage with it. 

Reviewer: Restore performance is the main consideration and should be explicitly called out.

Author: Will update spec to highlight this!

## Who this affects

- **Package consumers** get a way to discover, from the command line, which of their dependencies are seeking sponsorship — without having to visit nuget.org for every package individually.
- **Package authors** who've already added sponsorship links to their packages get more visibility for those links, since consumers can now see them as part of a workflow they already use (`dotnet list package`), not just on the package's web page.
- **Repository/org admins** get transparency into which of their organization's dependencies have funding needs, which can support internal conversations about sponsoring critical open-source dependencies.

## Explanation

This proposal adds `--sponsor` to `dotnet list package`, following the same opt-in report pattern as `--deprecated`, `--outdated`, and `--vulnerable`.

When you run `dotnet list package`, it restores the project and lists package dependencies:

```text
Top-level Package       Requested           Resolved
<PACKAGE_NAME>          <REQUESTED_VERSION> <RESOLVED_VERSION>
```

With `--sponsor`, output is grouped by project and includes sponsorship links for each matching package:

> ```text

Reviewer: Dotnet list package supports json output and we are missing a json sample in the spec.

Author: Will add tables and samples to visualize details

Project 'Contoso.App' has the following sponsorship information:
   > Contoso.Forms
      https://github.com/sponsors/contoso
      https://patreon.com/sponsor/contoso-forms
      https://opencollective.com/contoso-forms
   > Contoso.Utilities
      https://opencollective.com/contoso-utilities
```

When there are no packages looking for sponsors, there will be a message: 

```text

> No packages in project 'Contoso.Tools' are looking for sponsors.

Reviewer 2: You talked about a message for "your sources don't provide any sponsorship information". That should be called out here as well.

Author: Will add!

```

For private/internal feeds, users can explicitly query nuget.org sponsorship data:

```text

> dotnet list package --sponsor --source nuget.org

Reviewer 2: The sample should be the URI for the source. If the source were already configured to use a name, then users wouldn't need to specify the `--source` in the first place.

```suggestion
dotnet list package --sponsor --source https://api.nuget.org/v3/index.json
```

Author: Will update!

```

- **Scope:** solution-wide across all projects.
- **Transitive packages:** respects `--include-transitive`; no sponsor-specific behavior is required.

> - **Multiple sponsorship URLs:** up to 10 URLs are stored per package.

Reviewer 2: Storage is happening on the server already. I'm not clear what this means for client. Do we limit to 10 or we display whatever comes back from the server, etc?

Author: We would be displaying whatever comes back from the server. The reason it is 10 in the spec is because that is what is enforced on the server-side. I will update the spec to reflect this.

- **Empty state:** if a project has no sponsored packages, show a per-project message indicating none were found.

## Drawbacks

> - **No link priority/reorder support** With append/remove semantics, default ordering is oldest-added first. Not a v1 design need per this proposal's scope.

Reviewer 2: Order should probably just match what the server provides? nuget.org could change ordering or allow publishers to change the order in the future, right?

Author: Great point! I didn't consider that a future implementation could change this, I will update accordingly

> - **CLI-only scope for v1.** PM UI (Phase 2), popup link-out (Phase 3), Copilot/IDE, and Azure Search-backed surfaces are deferred.

Reviewer 2: Client focused reviewers may not be familiar with "Azure Search" terminology. I'd rephrase this in Client terms.

Author: Will update accordingly

> - **Freshness is cache-policy dependent.** Origin updates can be immediate, but user visibility depends on cache behavior.

Reviewer 2: Be specific about what cache this is referring to and the implications of that.

## Rationale and alternatives

Proposed solution: a `--sponsor` flag on `dotnet list package`, so developers can discover which dependencies are seeking sponsors.

Alternatives considered:

1. **Discovery alternatives**

> - Local cache or assets-file discovery: stale/non-authoritative for the current dependency graph — could show consumers outdated sponsorship information.

Reviewer 2: I don't really know what this is saying. I'd rewrite or remove it.

> - Separate tool outside `dotnet`/`nuget.exe`: inconsistent with existing CLI patterns consumers already know.

Reviewer 2: We discussed a potential new dotnet command, which you called out below. I'd just remove this. Not sure we even need to say "nuget.exe is out of scope". Others can chime in on that.

2. **Command alternatives**

> - New `dotnet sponsor` verb: higher cost than adding a report flag, and inconsistent with how consumers already discover `--deprecated`/`--outdated`/`--vulnerable` information.

Reviewer 2: This is great rationale for your design choice. I'd move this up under your "proposed solution" as the rationale for it.

Author: Will update!

- Combining sponsorship into an existing report flag: would blur `--sponsor`'s informational nature with report flags that call out actionable problems (deprecations, vulnerabilities), for limited benefit over a dedicated flag.

3. **Display alternatives**

> - JSON-only output: inconsistent with existing report UX that consumers already expect from `dotnet list package`.

Reviewer 2: Does this mean you're not supporting something like `--sponsor --format json`?

Author: This was intended as the default output rather than the proposed solution, I will update the spec to reflect that

> - Link-out popup flow: depends on separate nuget.org web work, deferred to Phase 3.

Reviewer 2: I know what this means but it's probably too vague for others to know what the popup is about. I also would say that's a future possibility with Phase 3, not necessarily an alternative proposal.

Author: Good call, will update accordingly

4. **Restore-time messaging**
- Restore is a noise-sensitive surface run on every build; An explicit, per-invocation `--sponsor` flag keeps this information fully opt-in without polluting the default build/restore loop.

**Impact of not doing this:** sponsorship links remain visible only on the nuget.org website; there is no way for `dotnet list package`, PM UI, or other client tooling to programmatically surface which dependencies are seeking sponsorship, limiting discoverability for package authors relying on this signal for funding.

## Prior Art

- **[`sponsor-link.md` (2023 community proposal)](https://github.com/NuGet/Home/blob/sponsor-link/proposed/2023/sponsor-link.md)**: closest precedent for the overall idea — introduced the `--sponsor` flag concept, sponsor metadata authors could set, a Visual Studio UI link, and an opt-in restore-time summary message.

> - **[`nuget/home#14739`](https://github.com/nuget/home/issues/14739)**: open customer ask for a Package Manager UI sponsorship button, confirming real demand for Phase 2.

Reviewer 2: I would focus on the Developer Community link https://developercommunity.visualstudio.com/t/Add-sponsorlink-visibility-in-the-NuGet/10456467 and just make tracking issue 14739 a sidenote here.

> - **`--deprecated`/`--vulnerable`**: precedent for opt-in report flags that consumers already use and understand.

Reviewer 2: highlight this more, it's a great point!

- **PM UI's existing project/license/report-abuse links**: precedent for Phase 2 — PM UI already shows package-supplied links to consumers using an established, low-risk pattern.

## Unresolved Questions

> 

Reviewer 2: nit/tip: I usually put something like "None at this time" when blank.

## Future Possibilities

- **Phase 2: VS Package Manager UI hyperlinks** for sponsorship links, directly addressing [nuget/home#14739](https://github.com/nuget/home/issues/14739).

> - **Phase 3: Compact link-out CLI display** using template-driven website links (`nuget.org/.../{id}/sponsor`) once sponsor popup URLs are independently addressable.

Reviewer 2: I'd file a nuget/home issue for this. We can set it to P3 for now.

> - **Azure Search-backed visibility** for PM UI Browse and nuget.org search experiences, helping consumers discover sponsorship needs while browsing for packages.

Reviewer: The NuGet client does not directly talk directly to any Azure Services. Not sure about this comment.

> - **Copilot/IDE surfaces**, per original customer interest cited in Motivation.

Reviewer 2: This is the MCP server, right? We have this issue about it https://github.com/nuget/home/issues/14738

Author: Yes! I will update the spec to include the issue.

```

In the case that a source either supports sponsorships but has no packages looking for them or a source does not support sponsorships, a respective message will be shown: 

```text
// No packages looking for sponsorship
No packages in project 'Contoso.Tools' are looking for sponsors.
```

```text
// Source does not provide sponsorship information
Source 'X' does not provide sponsorship information
```

For private/internal feeds, users can explicitly query nuget.org sponsorship data:

```text
dotnet list package --sponsor --source https://api.nuget.org/v3/index.json
```

`--sponsor` also supports JSON output via `--format json`, following the same base envelope as other report types (`version`, `parameters`, `sources`, `problems`). Since sponsorship is a per-package-ID concept independent of target framework, sponsored packages are listed per-project under `sponsoredPackages`, rather than nested under `frameworks` like `--deprecated`/`--outdated`/`--vulnerable`:

```json
{
  "version": 1,
  "parameters": "--sponsor",
  "sources": [
    "https://api.nuget.org/v3/index.json"
  ],
  "projects": [
    {
      "path": "/path/to/Contoso.App.csproj",

>       "sponsoredPackages": [

Reviewer 3: sponsoredPackages has a different connotation than what we have here. 

I think this is really sponsorablePackages or packages with sponsors.

Restore is also performance-sensitive: checking sponsorship requires reaching out over the network for every package, and doing so as part of restore would slow down builds and add repeated network calls to the frequent background restores that happen automatically while editing code (e.g., to keep IntelliSense/autocomplete up to date), not just when a developer explicitly builds.

## Who this affects

- **Package consumers** get a way to discover, from the command line, which of their dependencies are seeking sponsorship — without having to visit nuget.org for every package individually.
- **Package authors** who've already added sponsorship links to their packages get more visibility for those links, since consumers can now see them as part of a workflow (`dotnet package sponsor`), not just on the package's web page.
- **Repository/org admins** have an easier way to see sponsorship status alongside the other package information they already manage day-to-day (versions, licenses, vulnerabilities).

## Explanation

> This proposal adds a new `dotnet package sponsor` command, giving sponsorship discovery a dedicated and purpose-built surface. 

Reviewer 3: Consider providing a sample output of all sub commands.

When you run `dotnet package sponsor`, output is grouped by project and includes sponsorship links for each matching package:

```text
Project 'Contoso.App' has the following sponsorship information:
   > Contoso.Forms
      https://github.com/sponsors/contoso
      https://patreon.com/sponsor/contoso-forms
      https://opencollective.com/contoso-forms
   > Contoso.Utilities
      https://opencollective.com/contoso-utilities
```

> In the case that a source either supports sponsorships but has no packages looking for them or a source does not support sponsorships, a respective message will be shown: 

Reviewer 3: Sounds like we're leaning the direction of using the configured sources right?

If that's the case, to me the different form dotnet list package is much much smaller, so I'd be ok with `dotnet list package --sponsor` too.

More so than what we had discussed before.

- **[`npm fund`](https://docs.npmjs.com/cli/v10/commands/npm-fund/)**: `npm fund` provides precedent for this pattern. The `funding` field lives in package metadata, and npm's guidance suggests keeping funding links at the package or author level. 
npm notes that funding information can be noisy in the CLI and stale information could be problematic.
- **[Gallery UI Sponsorship Implementation](https://github.com/NuGet/Engineering/blob/prabora-needs-sponsorship-feature/Server.Specs/2025/NeedsSponsorship.md)**: companion server-side proposal for surfacing sponsorship needs directly in the nuget.org Gallery UI; relevant precedent/parallel effort for surfacing the same underlying sponsorship data.
- **`--deprecated`/`--vulnerable`**: precedent for opt-in report-style information that consumers already use and understand, informing this proposal's output/UX conventions even though it uses a dedicated verb rather than a flag.
- **PM UI's existing project/license/report-abuse links**: precedent for Phase 2 — PM UI already shows package-supplied links to consumers using an established, low-risk pattern.

> - **[`nuget/home#14739`](https://github.com/nuget/home/issues/14739)**: Open issue for PM UI sponsorship button

Reviewer: Could you also please add a link to the NuGet.org specification? (Maybe change the title to Prior Art / Related?)

Author: Will do! That said, it does live in the Server repo which is private so those who don't have permissions won't be able to access it

# **Surface Package Sponsorship Links to Developers**

- Author: [@kalebfik](https://github.com/kalebfik)
- GitHub Issue: [10703](https://github.com/NuGet/NuGetGallery/issues/10703)

## Summary

> Package maintainers can already add sponsorship links (GitHub Sponsors, Patreon, Open Collective, etc.) to their packages on nuget.org. Developers only see this today if they visit the package's nuget.org page. 

Reviewer 4: nitpick: This line has 2 sentences

This proposal brings that existing data into developer tooling — starting with a new `dotnet package list --sponsor` command — so developers can discover which dependencies are seeking sponsorship without leaving their workflow.

## Motivation

Sponsorship links already exist and are maintained by package owners on nuget.org.
The data is real and current, it's just invisible outside the website. 

This proposal adds a non-disruptive, opt-in way to surface sponsorship information, and is scoped as the first of three phases:

1. **Phase 1 (this proposal): CLI** — `dotnet package list --sponsor`.
2. **Phase 2: Visual Studio Package Manager UI** — surfacing sponsorship links using the same hyperlink pattern PM UI already uses for project/license/report-abuse URLs, directly addressing the demand in [nuget/home#14739](https://github.com/nuget/home/issues/14739).

> 3. **Phase 3: CLI popup link-out** — opening a nuget.org-hosted sponsorship page (`nuget.org/.../{id}/sponsor`) directly from the CLI, so developers get a richer, always-current destination without the CLI needing to hardcode or format every payment provider's link itself.

Reviewer 4: This line needs clarification.

1. I'm not sure if "CLI popup link-out" is supposed to be sufficiently clear just from the name, or if it's the proposed name for the feature. In case the name is intended to be self-describing, I can't guess what it means.
2. "opening a nuget.org-hosted .. page .. directly from the CLI". To me this sounds like it's proposing a text-based browser to get the nuget.org html and then render it directly in the CLI.
3. "so develpers get ...." to me, this sounds like a non-technical sales pitch which doesn't tell me anything about the features or implementation.

Author: I can see where it is unclear. 
The pop-up link is referring to the server-side implementation of this feature [(blog post for the feature)](https://devblogs.microsoft.com/dotnet/announcing-sponsorship-on-nugetdotorg-for-maintainer-appreciation/). 
I will update the spec to provide more clarity.

Providing sponsor links from nuget.org to the client is the key motivation for this proposal. 

> ### Why not restore?

Reviewer 4: pedantic: this seems like something that should be under rationale and alternaties, not motivation.

```text
dotnet package list --sponsor --source https://api.nuget.org/v3/index.json
```

`dotnet package list --sponsor` also supports JSON output via `--format json`. 
Since sponsorship is a per-package-ID concept independent of target framework, sponsorable packages are listed per-project under `sponsorablePackages`

```json
{
  "version": 1,
  "parameters": "--sponsor",
  "sources": [
    "https://api.nuget.org/v3/index.json"
  ],
  "projects": [
    {
      "path": "/path/to/Contoso.App.csproj",
      "sponsorablePackages": [
        {
          "id": "Contoso.Forms",
          "status": "available",
          "urls": [
            "https://github.com/sponsors/contoso",
            "https://patreon.com/sponsor/contoso-forms",
            "https://opencollective.com/contoso-forms"
          ]
        },
        {
          "id": "Contoso.Utilities",
          "status": "none",
          "urls": []
        },
        {
          "id": "Contoso.Internal.Shared",
          "status": "unsupported",
          "urls": []
        }
      ]
    }
  ]
}
```

### Distinguishing "No Sponsors" from "Source does not Provide Sponsorship Information" 

> For package sources implementing this proposal, the sponsorshipUrls field is expected to be present in registration documents. 

Reviewer 4: As per [the docs on the Package Metadata (registration) resource, there can be pages when there are many versions of a package](https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource#registration-pages-and-leaves).  For example, NuGet.Versioning: https://api.nuget.org/v3/registration5-gz-semver2/nuget.versioning/index.json

Will `sponsorshipUrls` be on just the root json, just the pages, or both?

Author: `sponsorshipUrls` would just be on the root json

If sponsorshipUrls is present and contains values, sponsorship information is displayed. 

If sponsorshipUrls is present but empty, the package is considered to have no sponsorship links.

If sponsorshipUrls is absent, the client treats sponsorship information as unavailable from that source. 

> **Sample registration index response using NewtonSoft.Json:**

Reviewer 4: I don't think there's any need to use a real package as an example. You can obfuscate it using something like Contoso.WebApi.Client

```json
{ 
  "@id": "https://api.nuget.org/v3/registration5-gz-semver2/newtonsoft.json/index.json",
  "@type": [
    "catalog:CatalogRoot", 
    "PackageRegistration", 
    "catalog:Permalink"
  ],
  "commitId": "afa91af1-9505-41b8-ad75-eab8e613db14",
  "commitTimeStamp": "2026-04-10T00:15:25.1492389+00:00",
  "count": 2,
  // ** Start of proposal ** //
  "sponsorshipUrls": [
    "https://github.com/sponsors/JamesNK"
  ],
  // ** End of proposal **//
  "items": [
    {
      "@id": "https://api.nuget.org/v3/registration5-gz-semver2/newtonsoft.json/index.json#page/3.5.8/12.0.1-beta2",
      "@type": "catalog:CatalogPage",
      "commitId": "afa91af1-9505-41b8-ad75-eab8e613db14",
      "commitTimeStamp": "2026-04-10T00:15:25.1492389+00:00",
      "count": 64,
      "items": [
        {
          "@id": "https://api.nuget.org/v3/registration5-gz-semver2/newtonsoft.json/3.5.8.json",
          "@type": "Package",
          "catalogEntry": {
            // ... additional catalog entry details
          }
        }
      ]
    }
  ]
}
```

- **Scope:** solution-wide across all projects.

> - **Transitive packages:** Included by default; users can use `--exclude-transitive` to exclude them.

Reviewer 4: Firstly, I think it's very reasonable that when asking what packages are looking for sponsors, to automatically include transitive packages by default. But thinking about it from the point of view of someone looking at `dotnet package list --help`, there's going to be both `--include-transitive` and `--exclude-transitive`. How will customers know when to use which? Which of the "modes" (outdated, vulnerable, sponsor, deprecated, or default list) are include by default and which are exclude by default?

Reviewer 2: One thought is does this option allow `--include-transitive false` ? If so, the default can be true as it is today and then for your command, if customers want only top-level they can provide false?

Author: Looking at the different modes, none of them have transitive packages showing by default. In this case, sponsorships would be the first (at least to my findings). One solution to users knowing that it includes transitive packages by default could be adding to --help signaling that they are on by default. Sponsorships would stand out, but I also think having to change the other flag descriptions to indicate whether they include/exclude transitive packages would be a bit much. 

I am open to other suggestions/solutions! This is just a floating idea specific to your question.

- **Multiple sponsorship URLs:** The CLI displays sponsorship URLs in the order received from the package source.
- **Empty state:** When a project has no sponsorable packages, display a per-project message indicating none were found.

## Drawbacks

> - **No link priority/reorder support:** The CLI will display sponsorship information in the same order as the queried package source, not performing any sorting of its own.

Reviewer 4: It's common for nuget.config files to list multiple package sources. Ideally just two, nuget.org and a company private feed, but it's also common for companies to have one feed per group. On one hand sponsorship is probably going to be nuget.org only, but if you're querying all package sources, and more than one of them return sponsorship info for a package (so, they're not using package source mapping to limit packages to a single source each), what order will they be displayed if the sponsor list is different?

Reviewer 2: For v1, should we hard-code a check for nuget.org, then in a future client, implement a merged metadata approach where other feeds could provide their own sponsorUrls?
It'd be ideal for v1 to merge, but at the same time, we have no evidence customers will implement sponsorship URLs on their own feeds.

Author: This is a good question. I am actively exploring how a separate source (in this case, Azure Artifacts) works with upstreaming registration information. There could be cases where that information is never reflected in other sources.

- **CLI-only scope:** PM UI (Phase 2), popup link-out (Phase 3), Copilot/IDE, and NuGet.org/PM UI Search surfaces are deferred.

## Rationale and Alternatives

Why `dotnet package list --sponsor`? Why not `dotnet package sponsor`?

```bash
dotnet package sponsor
```

This approach treats sponsorships similarly to npm's `npm fund` command. 

Benefits considered: 
- Clear intent and discoverability
- Provides room for future sponsorship-focused experiences. 
- Naturally supports interactive workflows if introduced in future implementations. 

**Selected Approach** 

```bash
dotnet package list --sponsor 
```

This was chosen because sponsorship information is currently treated as package metadata rather than a standalone workflow.

Rationale: 
- Aligns with existing package reporting patterns
- Sponsorhsip information is retrieved alongside other package metadata.
- Allows reuse of existing package discovery and reporting infrastructure

The current proposal focuses on displaying sponsorship information rather than providing sponsorship-specific actions. 

> 1. **Command alternatives**

Reviewer: I think this entire list (lines 219 to 231) can be removed. 

1. The fund naming can be mentioned in 191. Example:  Why not `dotnet package sponsor` or `dotnet package fund` ?
2. JSON-only output was never really an option, so let's remove it.
3. Restore-time messaging was already mentioned at the top of the document. @zivkan has a comment saying that it could be moved to this section. Either approach sounds good as long as it's not repeated in the doc.
4. We will not revisit the sponsorship information being baked into the package metadata as part of this document. That was covered as part of the initial specification and out of scope for this design, so we should remove this.
 
Impact of not doing this section can also be merged with Motivation at the top of this document.

```text
Added 342 packages in 4s

3 packages are looking for funding 
  run `npm fund` for details
```

After running `npm fund`, they receive this output: 

```text
my-project
├── https://github.com/sponsors/expressjs
│   └── express@4.21.2
├── https://opencollective.com/babel
│   └── @babel/core@7.28.0
└── https://github.com/sponsors/sindresorhus
    └── p-limit@7.1.1
```

### Characteristics of npm's approach

- Funding information is stored within package metadata.
- Funding information is distributed through the package ecosystem.
- Funding discovery has dedicated command surface (`npm fund`).

## Comparison with Proposed Approach

### Metadata Source

#### NPM

Funding information is directly embedded in package metadata and distributed throughout the npm ecosystem

#### NuGet

Sponsorship information is retrieved from the selected package source and surfaced through client tooling

### Command Surface

#### NPM

```bash
npm fund
```

Dedicated funding workflow. 

#### NuGet

```bash
dotnet package list --sponsor
```

Report-style experience aligned with existing commands.

## Why not mirror npm exactly?

> 1. Sponsorship information is treated as package metadata that's exposed by a package source instead of stored directly.

Reviewer 4: ```suggestion
1. Sponsorship information needs to be queried from the package source, rather than embedded in the package.
```

I think this is what this line means. I find the original wording confusing. Sponsorship links is metadata, it's not the package contents, so I feel like saying "is treated as package metadata" is a truism that adds words without clarifying. I also find "exposed by a package source instread of stored directly" to not be very clear, but that's definitely personal preference.

Reviewer 2: 👍

# **Surface Package Sponsorship Links to Developers**

- Author: [@kalebfik](https://github.com/kalebfik)
- GitHub Issue: [10703](https://github.com/NuGet/NuGetGallery/issues/10703)

## Summary

> This spec proposes adding a `--sponsor` flag to the exisitng `dotnet package list` command to expose sponsorship links in the CLI using the Registration API.

Reviewer 2: My understanding is this is a design spec, and the https://github.com/NuGet/Home/pull/14997 spec is a technical spec.
With that mindset, I'd expect the summary here to be more about the _why_ and less about the _how_.

## Motivation

> Package maintainers can already add sponsorship links (GitHub Sponsors, Patreon, Open Collective, etc.) to their packages on nuget.org. 

Reviewer 2: Link to the spec or blogpost integrating sponsors on nuget.org?
Example, https://learn.microsoft.com/nuget/nuget-org/package-sponsorship-on-nuget-org

Developers only see this today if they visit the package's nuget.org page. 

> This feature will assist in increasing discoverability for packages that need or are seeking funding. 

Reviewer 2: nit: wording?

```suggestion
Surfacing sponsorship data in the Client will assist in increasing discoverability for packages that need or are seeking funding. 
```

### Why not restore?

> Restore runs on every build; adding informational (non-actionable) content there, even opt-in, risks alarm fatigue and encourages developers to disable the feature entirely rather than engage with it. 

Reviewer 2: Is the feature capable of being disabled?

> Restore is also performance-sensitive: checking sponsorship requires reaching out over the network for every package, and doing so as part of restore would slow down builds and add repeated network calls to the frequent background restores that happen automatically while editing code (e.g., to keep IntelliSense/autocomplete up to date), not just when a developer explicitly builds.

Reviewer 2: nit: sentence is a bit lengthy. Condense?

When you run `dotnet package list --sponsor`, output is grouped by project and includes sponsorship links for each matching package:

**One Sponsorship Link** 

```text
Top-level Package        Sponsor
> Contoso.Tools          https://github.com/sponsors/username
```
**Ten Sponsorship Links** (matching NuGet.org's current policy cap)

```text
Top-level Package        Sponsor
> Contoso.Utility        https://domain1.com/user
                         https://domain1.com/user2
                         https://domain2.com/user
                         https://domain2.com/user2
                         https://domain2.com/user3
                         https://domain3.com/sponsors/user
                         https://domain3.com/sponsors/user2
                         https://domain4.com/123
                         https://domain4.com/456
                         https://domain4.com/789
```

Two distinct empty states are possible, and they are given distinct per-package markers so a user scanning only the table can distinguish "this package has no sponsors" with "this source can't tell us either way":

> - `(none)` — the source supports sponsorship data and confirms this package has no active sponsorship links.

Reviewer 2: I believe you've talked about only showing packages that have sponsors in the output.
Do we still need the `(none)` status in that case?

> - `(unsupported)` — the source does not provide sponsorship data at all, so no package from that source can be evaluated; this is a source-level limitation, not information about the package itself.

Reviewer 2: When would a package be in this scenario?
My thought is it would have to be part of a configuration where at least 1 package source supports sponsorship and 1 package source which does not. 
Then, this package was found on package source which does not support sponsorship.

I'm wondering if the idea of "just show packages that we found sponsors for" and "output a message indicating no package sources provide sponsorship data" is enough to cover the concern?

**Zero Sponsorship Links** (source supports sponsorship data; package has none)

```text
Top-level Package        Sponsor
> Contoso.Tools          (none)

> ```

Reviewer 2: Would we output this or would we just show a message? This implies that we are listing 100% of all packages and putting `(none)` in every Sponsor column.

**Source doesn't provide sponsorship information** 

```text
Top-level Package        Sponsor
> Contoso.Tools          (unsupported)

> 

Reviewer 2: See other feedback. Table/JSON output may not be needed for this scenario.

Source 'X' does not provide sponsorship data
```

> Users may explicity specify a package source using `--source`. 

Reviewer 2: We should mention that if Package Source Mapping is enabled, then this will fail unless the provided source is mapped (But in that case, there was no need to specify `--source` in the first place)

```text
dotnet package list --sponsor --source https://api.nuget.org/v3/index.json
```

`dotnet package list --sponsor` also supports JSON output via `--format json`. 
Since sponsorship is a per-package-ID concept independent of target framework, sponsorable packages are listed per-project under `sponsorablePackages`

> ```json

Reviewer 2: Let's show the exact command prior to each output.

### Distinguishing "No Sponsors" from "Source does not Provide Sponsorship Information" 

For package sources implementing this proposal, the sponsorshipUrls field is expected to be present in the registration index.

If sponsorshipUrls is present and contains values, sponsorship information is displayed. 

If sponsorshipUrls is present but empty, the package is considered to have no sponsorship links.

If sponsorshipUrls is absent, the client treats sponsorship information as unavailable from that source. 

## Drawbacks

- **No link priority/reorder support:** The CLI will display sponsorship information in the same order as the queried package source, not performing any sorting of its own.
- **CLI-only scope:** PM UI (Phase 2), popup link-out (Phase 3), Copilot/IDE, and NuGet.org/PM UI Search surfaces are deferred.

## Rationale and Alternatives

Why `dotnet package list --sponsor`? Why not `dotnet package sponsor` or `dotnet package fund`?

```bash
dotnet package sponsor
```

This approach treats sponsorships similarly to npm's `npm fund` command. 

Benefits considered: 
- Clear intent and discoverability
- Provides room for future sponsorship-focused experiences. 

> - Naturally supports interactive workflows if introduced in future implementations. 

Reviewer 2: I know you had considered the differences between `dotnet package list` and `dotnet package sponsor` and decided to focus on the filtering mechanism we have already in `dotnet package list`. This section captures that but the extra details here feel a bit subjective/distracting. I'd make it a little more focused.

Rationale: 
- For V1, the idea is to expose which packages have sponsorship information, similar to how `--vulnerable` exposes which packages have vulnerabilites. 
- Discoverability will happen when using `dotnet package list -help` when looking for the `--sponsor` flag
- Introducing sponsorships through a flag similar to other reporting patterns can allow us to analyze the feature and how it is used, before creating a dedicated command. 

**Selected Approach** 

```bash
dotnet package list --sponsor 
```

This was chosen because sponsorship information is currently treated as package metadata rather than a standalone workflow.

Rationale: 
- Aligns with existing package reporting patterns
- Sponsorhsip information is retrieved alongside other package metadata.

> - Allows reuse of existing package discovery and reporting infrastructure

Reviewer 2: Someone repeats the above. I'd consolidate this entire section.

The current proposal focuses on displaying sponsorship information rather than providing sponsorship-specific actions. 

> ### PMUI Integration

Reviewer 2: We could also have a separate spec for PM UI. The title of this spec is Sponsorship in CLI so it's probably not worth including PM UI details here.

> Endpoint Registartion: 

Reviewer 2: ```suggestion
Endpoint Registration: 
```

| Scenario | Endpoint |
|--------|--------|
| Browse/Search | Search |
| Install/Update/Details | Registration |

**Justification for Registration API** 

While using the Search API would be a easier technical implementation, installed-package workflows retrieve metadata through the Registration API. 
Having sponsorship information published through the Registration API would allow that data available without requiring additonal endpoint lookups.

Additionally, the command `dotnet package list` looks at packages that are installed and relies on Registration metadata. 

> Using the Search API would require a second retrieval path, while publishing sponsorships through Registration allows the command to retrieve sponsorship data alongside existing package metadata using the same data flow. 

Reviewer 2: This seems like technical detail? Also, I believe you're creating a dedicated spec for Registration API right?

> 

Reviewer 2: nit: blank lines

> **Impact of not doing this:** Sponsorship links remain visible only on the nuget.org website. There would be no way for the CLI, PM UI, or other client tooling to programmatically surface which dependencies are seeking sponsorship, limiting discoverability for package authors relying on this signal for funding.

Reviewer 2: Maybe move this to Motivation or another primary section?

## Prior Art/Related

- [**Companion server spec**](https://devdiv.visualstudio.com/DevDiv/_git/NuGet.Services/pullrequest/763096?_a=files&iteration=2&base=1): Proposes implementation for supporting PackageID level metadata in the Registration API.
- **[`npm fund`](https://docs.npmjs.com/cli/v10/commands/npm-fund/)**: `npm fund` provides precedent for this pattern. The `funding` field lives in package metadata, and npm's guidance suggests keeping funding links at the package or author level. 
npm notes that funding information can be noisy in the CLI and stale information could be problematic.
- **[Gallery UI Sponsorship Implementation](https://github.com/NuGet/Engineering/blob/prabora-needs-sponsorship-feature/Server.Specs/2025/NeedsSponsorship.md)**: companion server-side proposal for surfacing sponsorship needs directly in the nuget.org Gallery UI; relevant precedent/parallel effort for surfacing the same underlying sponsorship data.
- **`--deprecated`/`--vulnerable`**: precedent for opt-in report-style information that consumers already use and understand, informing this proposal's output/UX conventions even though it uses a dedicated verb rather than a flag.

> - **PM UI's existing project/license/report-abuse links**: precedent for Phase 2 — PM UI already shows package-supplied links to consumers using an established, low-risk pattern.

Reviewer 2: Good point here, but there's not much PM UI detail in the spec. If you decide not to separate the PM UI spec, I'd put more of this detail into the PM UI section above.

- **[`nuget/home#14739`](https://github.com/nuget/home/issues/14739)**: Open issue for PM UI sponsorship button

## Unresolved Questions

> None at this time

Reviewer 2: 👀

## Future Possibilities

While `dotnet package list --sponsor` is the proposed experience, future iterations could expand functionality:

**Enhanced CLI Experiences:**

> - Interactive sponsor selection experiences.

Reviewer 2: I'd be specific. This just means navigating the list of URLs and the ability to "Open in Browser", right?

> - Filtering, grouping, or prioritizing sponsorship opportunities.

Reviewer 2: ```suggestion
- Filtering, grouping, or prioritizing sponsorship ..(providers?)....
```

- Additional ecosystem engagement scenarios inspired by commands like `npm fund`.

- **Phase 2: VS Package Manager UI hyperlinks** for sponsorship links, directly addressing [nuget/home#14739](https://github.com/nuget/home/issues/14739).
- **Phase 3: Link to Package Specific Sponsor Page** using template-driven website link (`nuget.org/.../{id}/sponsor`) to reach the sponorship page specific to the package on nuget.org.

> - **Copilot/IDE surfaces**: See [Issue 14738](https://github.com/NuGet/Home/issues/14738)

Reviewer 2: I think #14738 is about AI Agents, not specifically Copilot and not related to IDEs.

## Appendix
NPM Fund Comparison

NPM provides a dedicated `npm fund` command that helps developers discover funding opportunities for packages they depend on.
Package authors specify funding information directly in their package metadata using the `funding` field. 

```json
{
  "funding": {
    "type": "individual",
    "url": "http://example/donate"
  },

  "funding": {
    "type": "patreon",
    "url": "https://domain/my-account"
  },
}
```

When a package containing funding metadata is installed, npm displays a summary message: 

```text
Added 342 packages in 4s

3 packages are looking for funding 
  run `npm fund` for details
```

After running `npm fund`, they receive this output: 

```text
my-project
├── https://domain/sponsors/user
│   └── express@4.21.2
├── https://domain/user
│   └── @babel/core@7.28.0
└── https://domain/sponsors/user
    └── p-limit@7.1.1
```

### Characteristics of npm's approach

- Funding information is stored within package metadata.
- Funding information is distributed through the package ecosystem.
- Funding discovery has dedicated command surface (`npm fund`).

## Comparison with Proposed Approach

### Metadata Source

> #### NPM

Reviewer 2: There may be too many section headings overall, but especially here. I'd only use a new heading level if there's going to be 2 distinct sections at that level.

Funding information is directly embedded in package metadata and distributed throughout the npm ecosystem

#### NuGet

Sponsorship information is retrieved from the selected package source and surfaced through client tooling

### Command Surface

#### NPM

```bash
npm fund
```

Dedicated funding workflow. 

#### NuGet

```bash
dotnet package list --sponsor
```

Report-style experience aligned with existing commands.

## Why not mirror npm exactly?

1. Sponsorship information needs to be queried from the package source, rather than embedded within the package. 
2. Sponsorship discovery follows existing NuGet resolution behavior 
3. Restore output remains unchanged to avoid additional noise.

> Users explicitly opt into sponsorship discovery instead of receiving sponsorship messaging during restore operations. 

Reviewer 2: Should this be the 4th bullet?

    }
  ],
  "projects": [
    {
      "path": "/path/to/Contoso.App.csproj",
      "sources": [
        {
          "source": "https://api.example.org/v3/index.json",
          "sponsorablePackages": []
        },
        {
          "source": "https://packages.contoso.com/v3/index.json",
          "sponsorablePackages": []
        }
      ]
    }
  ]
}
```

**Package Sources, Package Source Mapping (PSM), and `--source`**

The command selects the enabled package sources from NuGet configuration.
When PSM is disabled, every package ID is queried against every selected source.
For privacy reasons, PSM is honored for sponsorship reporting.
When PSM is enabled, each package ID is queried only against sources mapped to that package.

Specifying a source with `--source <SOURCE>` replaces the configured source selection but does not bypass PSM.
If a package is not mapped to a selected source, the package is not queried, and the console displays:

```text
Package `Contoso.Private` was not queried for sponsorship details because none of the selected package sources match its Package Source Mapping.

> ```

Reviewer 2: Left feedback around this on the technical spec, which should be changed here as well.

Currently, nuget.org is the only source that exposes sponsorship information through its Registration response.

> Other sources will be queried but will produce empty results until they implement the same contract.

Reviewer 2: We have been discussing an improvement on this idea, so reminder to change this when that's clarified.

> Sponsorship details from an upstream source are not shown unless that source exposes the sponsorship details through its own Registration response.

Reviewer 2: ```suggestion
Sponsorship details from an Azure Artifacts upstream source are not shown unless that source exposes the sponsorship details through its own Registration response.
```

For the initial implementation of this feature, nuget.org will be the only source used to retrieve sponsorship information.

> While the server design is intended to be extensible to other feeds, we don't know when or if other feeds will implement sponsorship. To avoid complicating the client logic, results from multiple feeds won't be merged.

Reviewer 2: We no longer need to merge, as we are just listing per-source.
Additionally, we may have a way to let sources declare if they support id-level metadata.

# **Proposal to Surface Sponsorship Information in the CLI**

- Author: [@kalebfik](https://github.com/kalebfik)
- GitHub Issue: [10703](https://github.com/NuGet/NuGetGallery/issues/10703)

> - [Accompanying Technical Implementation Spec](https://github.com/NuGet/Home/blob/dev-kalebfika-sponsorTechSpec/accepted/2026/implementing-sponsorship-in-cli.md)

Reviewer 2: I'd just remove this reference. Once merged, this branch will be deleted and the link will become invalid.

**Example registration root entry with `metadata` and `sponsorshipUrls` added:**

```jsonc
{
  "@id": "https://api.nuget.org/v3/registration5-gz-semver2/contoso.webapi.client/index.json",
  "@type": [
    "catalog:CatalogRoot",
    "PackageRegistration",
    "catalog:Permalink"
  ],
  "commitId": "afa91af1-9505-41b8-ad75-eab8e613db14",
  "commitTimeStamp": "2026-04-10T00:15:25.1492389+00:00",
  "count": 2,
  // ** Start of proposal ** //
  "metadata": {
    "sponsorshipUrls": [
      "https://github.com/sponsors/contoso"
    ]
  }
  // ** End of proposal **//
}
```

- **Scope:** solution-wide across all projects.
- **Transitive packages:** Included by default; when using `dotnet package list --help`, the description will state that transitive packages are included by default.
- **Multiple sponsorship URLs:** The CLI displays sponsorship URLs in the order received from the package source.
- **nuget.org URL limit:** nuget.org currently enforces a maximum of 10 URLs per package.
- **Empty state:** A missing, null, or empty `sponsorshipUrls` value is treated as a successful empty result.

> A detailed implementation proposal can be found [here](https://github.com/NuGet/Home/blob/dev-kalebfika-sponsorTechSpec/accepted/2026/sponsorship-CLI-Spec.md).

Reviewer 2: Same as above, I'd remove this.
